### ISSUE TITLE
Joins lose columns when using rsqlite (reprex)

### ISSUE THREAD

**Issue by JohnMount** [tidyverse/dplyr#2823]
Joins lose columns when using rsqlite.  reprex below comparing result on data.frame to sqlite results.

``` r
suppressPackageStartupMessages(library('dplyr'))

runJoinExperiment <- function(prefix, sc, eagerCompute) {
  names <- paste('t', prefix, 1:10, sep= '_')
  joined <- NULL
  for(ni in names) {
    di <- data.frame(x=1:3, y=1:3)
    if(!is.null(sc)) {
      ti <- copy_to(sc, di, ni)
    } else {
      ti <- di
    }
    if('NULL' %in% class(joined)) {
      joined <- ti
    } else {
      joined <- left_join(joined, ti, by= 'x')
      if(eagerCompute) {
        joined <- compute(joined)
      }
    }
  }
  compute(joined)
}

# works as expected
runJoinExperiment('inmem', NULL, FALSE)
#>   x y.x y.y y.x.x y.y.y y.x.x.x y.y.y.y y.x.x.x.x y.y.y.y.y y.x.x.x.x.x
#> 1 1   1   1     1     1       1       1         1         1           1
#> 2 2   2   2     2     2       2       2         2         2           2
#> 3 3   3   3     3     3       3       3         3         3           3
#>   y.y.y.y.y.y
#> 1           1
#> 2           2
#> 3           3

# throws
sc <- src_sqlite(":memory:", create = TRUE)

tryCatch(
  runJoinExperiment('sqlitea', sc, FALSE),
  error = function(e) print(e)
)
#> <Rcpp::exception in rsqlite_send_query(conn@ptr, statement): parser stack overflow>

# incorrect result (missing columns)
runJoinExperiment('sqliteb', sc, TRUE)
#> Source:   query [?? x 3]
#> Database: sqlite 3.11.1 [:memory:]
#> 
#> # A tibble: ?? x 3
#>       x   y.x   y.y
#>   <int> <int> <int>
#> 1     1     1     1
#> 2     2     2     2
#> 3     3     3     3

packageVersion("dplyr")
#> [1] '0.5.0'
if(requireNamespace("dbplyr", quietly = TRUE)) {
  packageVersion("dbplyr")
}
if(requireNamespace("RSQLite", quietly = TRUE)) {
  packageVersion("RSQLite")
}
#> [1] '1.1.2'
R.Version()$version.string
#> [1] "R version 3.4.0 (2017-04-21)"
```

**Comment by krlmlr** [comment]
Thanks. I remember dplyr 0.5.0 had this problem, could you please retry the development version of dplyr?

**Comment by krlmlr** [comment]
Closing as duplicate of https://github.com/tidyverse/dplyr/issues/2823, it's unlikely to be an RSQLite problem.

**Comment by krlmlr** [comment]
Thanks, replicated with current dbplyr. Shorter reprex:

``` r
(df <- dplyr::left_join(
  dplyr::tibble(a = 1, b.x = 1, b = 1),
  dplyr::tibble(a = 1, b = 1),
  by = "a"
))
#> # A tibble: 1 x 4
#>       a   b.x b.x.x   b.y
#>   <dbl> <dbl> <dbl> <dbl>
#> 1     1     1     1     1
(sql <- dplyr::left_join(
  dbplyr::memdb_frame(a = 1, b.x = 1, b = 1),
  dbplyr::memdb_frame(a = 1, b = 1),
  by = "a"
))
#> # Source:   lazy query [?? x 4]
#> # Database: sqlite 3.11.1 [:memory:]
#>       a   b.x `b.x:1`   b.y
#>   <dbl> <dbl>   <dbl> <dbl>
#> 1     1     1       1     1
dplyr::compute(sql)
#> Error: Data source must be a dictionary
```

Removed previous comments to clean up thread.

**Comment by hadley** [comment]
The SQL generation looks ok. I think the problem is in `op_vars()` and occurs when one of the disambiguation prefixes matches the name of an existing column.

**Comment by hadley** [comment]
More minimal reprex:

```R
join_vars(c("a", "b.x", "b"), c("a", "b"), type = "left", by = list(x = "a", y = "b"))
#> $alias
#> [1] "a.x" "b.x" "b"   "a.y"
#> 
#> $x
#> [1] "a"   "b.x" "b"   NA   
#> 
#> $y
#> [1] NA  NA  NA  "a"
```

**Comment by hadley** [comment]
Fixed in bf0032b7a63cf81e97856d77a4c24c2e5335a076

**Comment by JohnMount** [comment]
Looks fixed, but RSQLite is issue a lot of warnings in the original larger examples.  I have cut it down to a reprex and will file.

``` r
suppressMessages(library("dplyr"))
packageVersion("dplyr")
#> [1] '0.7.1'

db <- DBI::dbConnect(RSQLite::SQLite(), 
                     ":memory:")
dR <- dplyr::copy_to(db,
                     data.frame(x = 3.077), 
                     'dR')
compute(left_join(dR, dR, by='x'))
#> Warning in rsqlite_fetch(res@ptr, n = n): Don't need to call dbFetch() for
#> statements, only for queries
#> # Source:   table<kzlkrvgirz> [?? x 1]
#> # Database: sqlite 3.19.3 [:memory:]
#>       x
#>   <dbl>
#> 1 3.077
```
